pulsar/consumer_partition.go

```diff
@@ -0,0 +1,508 @@
+				Message:  messages[0],
+			}
+			messageCh = pc.messageCh
+		} else {
+			// we are ready for more messages
+			queueCh = pc.queueCh
+		}
+
+		select {
+		case <-pc.closeCh:
+			return
+
+		case _, ok := <-pc.connectedCh:
+			if !ok {
+				return
+			}
+			pc.log.Debug("dispatcher received connection event")
+
+			// drain messages
+			messages = nil
+
+			// drain the message queue on any new connection by sending a
+			// special nil message to the channel so we know when to stop dropping messages
+			go func() {
+				pc.queueCh <- nil
+			}()
+			for m := range pc.queueCh {
+				// the queue has been drained
+				if m == nil {
+					break
+				}
+			}
+
+			// reset available permits
+			pc.availablePermits = 0
+			initialPermits := uint32(pc.queueSize)
+
+			pc.log.Debugf("dispatcher requesting initial permits=%d", initialPermits)
+			// send initial permits
+			if err := pc.internalFlow(initialPermits); err != nil {
+				pc.log.WithError(err).Error("unable to send initial permits to broker")
+			}
+
+		case msgs, ok := <-queueCh:
+			if !ok {
+				return
+			}
+			// we only read messages here after the consumer has processed all messages
+			// in the previous batch
+			messages = msgs
+
+		// if the messageCh is nil or the messageCh is full this will not be selected
+		case messageCh <- nextMessage:
+			// allow this message to be garbage collected
+			messages[0] = nil
+			messages = messages[1:]
+
+			// TODO implement a better flow controller
+			// send more permits if needed
+			pc.availablePermits++
+			flowThreshold := int32(math.Max(float64(pc.queueSize/2), 1))
+			if pc.availablePermits >= flowThreshold {
+				availablePermits := pc.availablePermits
+				requestedPermits := availablePermits
+				pc.availablePermits = 0
+
+				pc.log.Debugf("requesting more permits=%d available=%d", requestedPermits, availablePermits)
+				if err := pc.internalFlow(uint32(requestedPermits)); err != nil {
+					pc.log.WithError(err).Error("unable to send permits")
+				}
+			}
+		}
+	}
+}
+
+type ackRequest struct {
+	doneCh chan struct{}
+	msgID  MessageID
+	err    error
+}
+
+type unsubscribeRequest struct {
+	doneCh chan struct{}
+	err    error
+}
+
+type closeRequest struct {
+	doneCh chan struct{}
+	err    error
+}
+
+func (pc *partitionConsumer) runEventsLoop() {
+	defer func() {
+		pc.log.Infof("consumer=%d exiting events loop", pc.consumerID)
+	}()
+	for {
+		select {
+		case <-pc.closeCh:
+			return
+		case i := <-pc.eventsCh:
+			switch v := i.(type) {
+			case *ackRequest:
+				pc.internalAck(v)
+			case *unsubscribeRequest:
+				pc.internalUnsubscribe(v)
+			case *connectionClosed:
+				pc.reconnectToBroker()
+			case *closeRequest:
+				pc.internalClose(v)
+				return
+			}
+		}
+	}
+}
+
+func (pc *partitionConsumer) internalClose(req *closeRequest) {
+	defer close(req.doneCh)
+	if pc.state != consumerReady {
+		return
+	}
+
+	pc.state = consumerClosing
+	pc.log.Infof("Closing consumer=%d", pc.consumerID)
```

Comment:
Since we're repeating consumer Id, we should put the field directly in the logger, to ensure consistency. 

In this case though, the consumer id does not make any sense to the application, rather it's an internal id and I'd prefer to actually just hide it (for other purposes than client lib debug).